packages/protocol/contracts/governance/ReleaseGoldInstance.sol

```diff
@@ -0,0 +1,579 @@
+   */
+  function refundAndFinalize() external nonReentrant onlyReleaseOwnerAndRevoked {
+    require(getRemainingLockedBalance() == 0, "Total gold balanace must be unlocked");
+    uint256 beneficiaryAmount = revocationInfo.releasedBalanceAtRevoke.sub(totalWithdrawn);
+    require(
+      getGoldToken().transfer(beneficiary, beneficiaryAmount),
+      "Transfer of gold to beneficiary failed"
+    );
+    uint256 revokerAmount = getRemainingUnlockedBalance();
+    require(
+      getGoldToken().transfer(releaseOwner, revokerAmount),
+      "Transfer of gold to releaseOwner failed"
+    );
+    selfdestruct(releaseOwner);
+  }
+
+  /**
+   * @notice Revoke the future release schedule.
+   */
+  function revoke() external nonReentrant onlyReleaseOwner onlyRevocable {
+    require(!isRevoked(), "Release schedule instance must not already be revoked");
+    revocationInfo.revokeTime = block.timestamp;
+    revocationInfo.releasedBalanceAtRevoke = getCurrentReleasedTotalAmount();
+    emit ReleaseScheduleRevoked(
+      beneficiary,
+      revocationInfo.revokeTime,
+      revocationInfo.releasedBalanceAtRevoke
+    );
+  }
+
+  /**
+   * @notice Calculates the total balance of the release schedule instance including withdrawals.
+   * @return The total released instance gold balance.
+   * @dev The returned amount may vary over time due to locked gold rewards.
+   */
+  function getTotalBalance() public view returns (uint256) {
+    return getRemainingUnlockedBalance().add(getRemainingLockedBalance()).add(totalWithdrawn);
+  }
+
+  /**
+   * @notice Calculates the sum of locked and unlocked gold in the release schedule instance.
+   * @return The remaining total released instance gold balance.
+   * @dev The returned amount may vary over time due to locked gold rewards.
+   */
+  function getRemainingTotalBalance() public view returns (uint256) {
+    return getRemainingUnlockedBalance().add(getRemainingLockedBalance());
+  }
+
+  /**
+   * @notice Calculates remaining unlocked gold balance in the release schedule instance.
+   * @return The available unlocked release schedule instance gold balance.
+   */
+  function getRemainingUnlockedBalance() public view returns (uint256) {
+    return address(this).balance;
+  }
+
+  /**
+   * @notice Calculates remaining locked gold balance in the release schedule instance.
+   * @return The remaining locked gold of the release schedule instance.
+   * @dev The returned amount may vary over time due to locked gold rewards.
+   */
+  function getRemainingLockedBalance() public view returns (uint256) {
+    return getLockedGold().getAccountTotalLockedGold(address(this));
+  }
+
+  /**
+   * @dev Calculates the total amount that has already released up to now.
+   * @return The already released amount up to the point of call.
+   * @dev The returned amount may vary over time due to locked gold rewards.
+   */
+  function getCurrentReleasedTotalAmount() public view returns (uint256) {
```

Comment:
Should this be the function that checks the liquidity provision, rather than `withdraw`? Or in other words, when an instance is revoked when the liquidity provision has yet to be hit, what should happen?

Reply:
I added the check here, which is correct and something I overlooked.
I'm also keeping the check in `withdraw` since I think it makes sense for withdraw to revert with a message about liquidity provision, rather than reverting for the generic "you have 0 funds available to withdraw".